### ISSUE TITLE
Add spec entry points that don't trigger user code

### ISSUE THREAD

**Issue by jorendorff** [domenic/promises-unwrapping#73]
At the moment, the Loader draft spec is using these APIs:
- `new Promise`
- `Promise.resolve`
- `Promise.all`
- `Promise.prototype.then`
- `Promise.prototype.catch`

We need equivalents that users can’t sabotage by messing with `Promise.@@create` and `Promise.prototype`. Most of the promises we create are implementation details and users should not be able to observe their existence, much less disrupt their behavior.

@dherman was going to ask in the committee about how to spec this, but it'd be nice to just present something that works.

**Comment by erights** [comment]
I am so glad to hear this. It is great news that we'll need support for high integrity promises asap. I was very nervous that if we were only trying to pave the way for high integrity later, that we might get things wrong.

I suggest you do what SES will need to do: create a separate realm with normal promises, and then as-if run an initialization script that freezes the primordials. The loader's promises would then be from that realm.

**Comment by domenic** [comment]
@allenwb, I thought there was some "use the original value of X" in the spec, but I cannot find anything now.

**Comment by allenwb** [comment]
I don't think that this is the sort of design issue that needs to be taken to the committee, unless some controversy develops.

Those APIs should be easy, it's just a matter of defining abstract operations that encapsulate the specification of the functionality you need to use.  If it is the same as the (part) of the spec. of a user callable method then typically we would refactor so that the method specification also calls the abstract operation. 

I need to see your actual use cases to be sure, but it isn't clear to me that you will even need to use the object level promise abstraction for internal coordination within the built-in default loader mechanism.  We may find that a more straightforward approach is simply to provide an API for scheduling a "invoke abstract operation" micro-task and perhaps a few other operations at that level of scheduling.  Then we can specify asynchronous behaviors without having to define JS level closures or other non-observable objects, or worrying about integrity issues.

**Comment by allenwb** [comment]
@domenic I have pretty much purged all of that language from the spec.  In some cases it is replaced with named references to intrinsic, eg %ArrayPrototype%.  But in many cases it has been replaced with direct references to the appropriate abstract operations.

**Comment by jorendorff** [comment]
@domenic For example, evaluating an ArrayLiteral, like `[1, 2]`, is no longer specified in terms of `new Array` or `%Array%.[[Construct]]()` or anything like that. Instead, it calls [ArrayCreate](http://people.mozilla.org/~jorendorff/es6-draft.html#sec-arraycreate) which is like `new Array` but doesn’t invoke `%Array%[@@create]`.

**Comment by jorendorff** [comment]
@allenwb It is possible to specify the loader with or without using Promises internally. I've implemented it both ways now (in JS, not in EcmaSpeak).

Using Promises is more consistent and reads better, because the API is all asynchronous using Promises anyway. It's also less code—but that is partly because JS has pretty nice support for closures (EcmaSpeak sort of doesn’t).

**Comment by domenic** [comment]
@jorendorff I ran out of time to do this today, but you can anticipate the following, maybe tomorrow:
- `new Promise` -> `PromiseCreate(resolver)`
- `Promise.resolve` -> `PromiseFor(x)`
- `Promise.prototype.then` -> `PromiseThen(promise, onFulfilled, onRejected)`

`Promise.prototype.catch` just consists of passing in `undefined` as `onFulfilled` when calling `PromiseThen`.

Also, it looks like you don't need `CastToPromise` anymore, so I'll put that back inside `Promise.cast`.

`Promise.all` seems like a special case. First, I'm betting you'll want it to take a List, not an iterable. Next, if you want it to be truly unobservable, you'll probably be able to guarantee it never operates on anything but spec-created-promises, right? And, you'll want it to not call their `then` methods, which might be tampered with. So this all points to a rather simplified abstract operation that makes several assumptions and is also more magical in certain ways. It's going to be a tricky one to spec; might not be able to get to it this weekend.

Actually, now I'm a bit afraid some of these concerns apply to `Promise.prototype.then` as well. Hmm. We'll see.

**Comment by jorendorff** [comment]
Right now I'm using PromiseThen more or less exactly as proposed, because it practically always wants closures, which is to say, anonymous functions.

But you're right, I don't really know if that's the most sensible thing and we'll see.

**Comment by jorendorff** [comment]
I could get rid of Loader.js's use of Promise.all and replace it with a counter. I don't need the results, I just need to do more work when all the Promises are resolved, and propagate the error if any fail. But that's most of the work Promise.all does. Hmm.

**Comment by domenic** [comment]
> But that's most of the work Promise.all does. 

I don't think so. With the simplifying assumptions we have so far (List, not potentially exception-ey iterable; only-spec-promises, not thenables or scalars; results not needed; does not need to be generic in `this`) it'd be a much simpler function. One of us can spec it and see how it goes.

**Comment by domenic** [comment]
@jorendorff per recent discussions it sounds like this may be less necessary?

**Comment by jorendorff** [comment]
This is still necessary. CastToPromise is the thing I'm no longer using.
